api/mfa.go

```diff
@@ -265,8 +277,76 @@ func (a *API) ChallengeFactor(w http.ResponseWriter, r *http.Request) error {
 	return sendJSON(w, http.StatusOK, &ChallengeFactorResponse{
 		ID:           challenge.ID,
 		CreatedAt:    creationTime.String(),
-		ExpiresAt:    creationTime.Add(time.Second * challengeExpiryDuration).String(),
+		ExpiresAt:    creationTime.Add(time.Second * time.Duration(config.MFA.ChallengeExpiryDuration)).String(),
 		FactorID:     factor.ID,
 		FriendlyName: factor.FriendlyName,
 	})
 }
+
+func (a *API) VerifyFactor(w http.ResponseWriter, r *http.Request) error {
+	var err error
+	ctx := r.Context()
+	config := a.getConfig(ctx)
+	user := getUser(ctx)
+	instanceID := getInstanceID(ctx)
+	if !user.MFAEnabled {
+		return forbiddenError(MFANotEnabledMsg)
+	}
+
+	params := &VerifyFactorParams{}
+	jsonDecoder := json.NewDecoder(r.Body)
+	err = jsonDecoder.Decode(params)
+	if err != nil {
+		return badRequestError("Please check the params passed into VerifyFactor: %v", err)
+	}
+
+	factor, err := models.FindFactorByChallengeID(a.db, params.ChallengeID)
+	if err != nil {
+		if models.IsNotFoundError(err) {
+			return notFoundError(err.Error())
+		}
+		return internalServerError("Database error finding factor").WithInternalError(err)
+	}
+
+	challenge, err := models.FindChallengeByChallengeID(a.db, params.ChallengeID)
+	if err != nil {
+		if models.IsNotFoundError(err) {
+			return notFoundError(err.Error())
+		}
+		return internalServerError("Database error finding Challenge").WithInternalError(err)
+
+	}
+	hasExpired := time.Now().After(challenge.CreatedAt.Add(time.Second * time.Duration(config.MFA.ChallengeExpiryDuration)))
+	if hasExpired {
```

Comment:
if the challenge has expired, do we want to do a clean up and delete it from the db?

Reply:
Hmm yeah sure let's do that. Update this in a bit